app/slave/cluster_slave.py

```diff
@@ -74,24 +75,36 @@ def _start_heartbeat(self):
 
     def _run_heartbeat(self):
         try:
-            self._send_heartbeat_to_master()
-            self._heartbeat_failure_count = 0
+            response = self._send_heartbeat_to_master()
+
+            # A non HTTP 200 response indicates that the master is up, but does not recognize the slave.
+            # The slave tries to reconnect to master at this point.
+            # This is also counted as a heartbeat failure and the slave takes an action accordingly
+            if response.status_code != http.client.OK:
```

Comment:
`!=200` might be a little strict
`if response.status_code > 299`

Reply:
`response.ok`?